File: examples/python/graph_example.py
"""MembrainGraph -- knowledge graph with multi-hop traversal and persistence."""

import random

from membrain import MembrainClient, MembrainGraph


def embed(seed: float, dimension: int = 128) -> list[float]:
    random.seed(seed)
    return [random.gauss(0, 1) for _ in range(dimension)]


def main() -> None:
    with MembrainClient() as client, MembrainGraph() as graph:
        topics = [
            ("Python is a high-level programming language", 0.95),
            ("Rust provides memory safety without garbage collection", 0.95),
            ("HNSW is a graph-based ANN algorithm", 0.90),
            ("Vector databases store high-dimensional embeddings", 0.90),
            ("RAG combines retrieval with generation", 0.85),
        ]

        for i, (statement, confidence) in enumerate(topics):
            result = client.store_fact(statement, confidence=confidence)
            if result.success and result.id:
                graph.add_node(result.id, embed(seed=float(i)), confidence=confidence)

        print(f"Graph: {graph.node_count()} nodes, {graph.edge_count()} edges")

        query = embed(seed=2.5)
        result = graph.query(query, max_hops=3, top_k=3)
        print(f"\nMulti-hop query: {result.hops_performed} hops, {result.nodes_visited} visited")
        for node in result.nodes:
            memory = client.get(node.memory_id)
            print(f"  {memory.content[:60]}... (score={node.score:.3f}, hop={node.hop_distance})")

        serialized = graph.save()
        print(f"\nSerialized: {len(serialized)} bytes")

    restored = MembrainGraph.load(serialized)
    print(f"Restored: {restored.node_count()} nodes, {restored.edge_count()} edges")
    restored.close()


if __name__ == "__main__":
    main()

File: examples/javascript/graph_example.mjs
/** MembrainGraph -- knowledge graph with multi-hop traversal and persistence. */

import { MembrainClient, MembrainGraph } from "membrain";

const DIMENSION = 128;

function embed(seed) {
  const out = [];
  let state = seed * 1000 + 1;
  for (let i = 0; i < DIMENSION; i++) {
    state = (state * 1103515245 + 12345) & 0x7fffffff;
    out.push((state / 0x7fffffff) * 2 - 1);
  }
  return out;
}

const client = new MembrainClient();
const graph = new MembrainGraph();

try {
  const topics = [
    ["Python is a high-level programming language", 0.95],
    ["Rust provides memory safety without garbage collection", 0.95],
    ["HNSW is a graph-based ANN algorithm", 0.9],
    ["Vector databases store high-dimensional embeddings", 0.9],
    ["RAG combines retrieval with generation", 0.85],
  ];

  for (let i = 0; i < topics.length; i++) {
    const [statement, confidence] = topics[i];
    const result = client.storeFact(statement, confidence);
    if (result.success && result.id) {
      graph.addNode(result.id, embed(i), confidence);
    }
  }

  console.log(`Graph: ${graph.nodeCount()} nodes, ${graph.edgeCount()} edges`);

  const query = embed(2.5);
  const result = graph.query(query, 3, 3);
  console.log(`\nMulti-hop query: ${result.hops_performed} hops, ${result.nodes_visited} visited`);
  for (const node of result.nodes) {
    const memory = client.get(node.memory_id);
    console.log(`  ${memory.content.slice(0, 60)}... (score=${node.score.toFixed(3)}, hop=${node.hop_distance})`);
  }

  const serialized = graph.save();
  console.log(`\nSerialized: ${serialized.length} bytes`);

  const restored = MembrainGraph.load(serialized);
  console.log(`Restored: ${restored.nodeCount()} nodes, ${restored.edgeCount()} edges`);
  restored.close();
} finally {
  graph.close();
  client.close();
}
